unstructured/ingest/connector/pinecone.py

```diff
@@ -0,0 +1,125 @@
+        return index
+
+    def create_session_handle(self) -> PineconeSessionHandle:
+        service = self.create_pinecone_object(self.api_key, self.index_name, self.environment)
+        return PineconeSessionHandle(service=service)
+
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def upsert_batch(self, batch):
+        import pinecone.core.client.exceptions
+
+        index = self.session_handle.service
+        try:
+            response = index.upsert(batch)
+        except pinecone.core.client.exceptions.ApiException as api_error:
+            raise WriteError(f"http error: {api_error}") from api_error
+        logger.debug(f"results: {response}")
+
+
+@dataclass
+class SimplePineconeConfig(BaseConnectorConfig):
+    api_key: str
+    index_name: str
+    environment: str
+
+
+@dataclass
+class PineconeDestinationConnector(BaseDestinationConnector):
+    write_config: WriteConfig
+    connector_config: SimplePineconeConfig
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["pinecone"], extras="pinecone")
+    def initialize(self):
+        pass
+
+    def write_dict(self, *args, dict_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        logger.info(
+            f"Inserting / updating {len(dict_list)} documents to destination "
+            f"index at {self.connector_config.index_name}",
+        )
+
+        # this is advised to be 100 at maximum in pinecone docs, however when we
+        # chunk content, we hit to the object size limits, so we decrease the batch
+        # size even more here
+        pinecone_batch_size = 10
+
+        num_processes = 1
+        if num_processes == 1:
+            for i in range(0, len(dict_list), pinecone_batch_size):
+                self.write_config.upsert_batch(dict_list[i : i + pinecone_batch_size])  # noqa: E203
+
+        else:
+            with mp.Pool(
+                processes=num_processes,
+            ) as pool:
+                pool.map(
+                    self.write_config.upsert_batch,
+                    [
+                        dict_list[i : i + pinecone_batch_size]  # noqa: E203
+                        for i in range(0, len(dict_list), pinecone_batch_size)
+                    ],  # noqa: E203
+                )
+
+    def write(self, docs: t.List[BaseIngestDoc]) -> None:
+        dict_list: t.List[t.Dict[str, t.Any]] = []
+        for doc in docs:
+            local_path = doc._output_filename
+            with open(local_path) as json_file:
+                dict_content = json.load(json_file)
+
+                # assign element_id and embeddings to "id" and "values"
+                # assign everything else to "metadata" field
+                dict_content = [
+                    {
+                        "id": element.pop("element_id", None),
+                        "values": element.pop("embeddings", None),
+                        "metadata": {k: json.dumps(v) for k, v in element.items()},
```

Comment:
This last part creates a big long string of metadata. Are we sure thats what we want to do?

Looks like this:

Reply:
Agreed, addressed with [pass metadata fields individually](https://github.com/Unstructured-IO/unstructured/pull/1774/commits/f2786ce167d91a69f69f28e50f8a2a1f1e0b146c)